cameo/flux_analysis/analysis.py

```diff
@@ -367,6 +378,100 @@ class _PhenotypicPhasePlaneChunkEvaluator(object):
     def __init__(self, model, variable_reactions):
         self.model = model
         self.variable_reactions = variable_reactions
+        objective_reactions = [reaction for reaction in model.reactions
+                               if reaction.objective_coefficient != 0]
+        if len(objective_reactions) != 1:
+            raise Exception('multiple objectives not supported')
+        self.product_reaction = objective_reactions[0]
+        self.c_source_reactions = self.get_c_source_reactions()
+
+    def get_c_source_reactions(self):
+        """ carbon source reactions
+
+        Returns
+        -------
+        list
+           list of media reaction currently with negative flux or 0 if non defined
+        """
+        medium_reactions = [self.model.reactions.get_by_id(reaction) for reaction in self.model.medium.reaction_id]
+        self.model.solve()
+        source_reactions = [reaction for reaction in medium_reactions if
+                            (reaction.n_carbon() > 0) and (reaction.flux < 0)]
+        if len(source_reactions) == 0:
+            return 0
+        return source_reactions
+
+    @classmethod
+    def total_flux(cls, reactions, by_carbon=True):
+        """ sum flux for reactions
+
+        Parameters
+        ----------
+        reactions : list
+            list of reactions the reactions to sum flux for
+
+        by_carbon : bool
+            return flux of carbon instead of summed metabolites
+
+        Returns
+        -------
+        float
+            summed flux for a set of reactions. Possibly the flux of carbon """
+        flux = 0
+        for rxn in reactions:
+            carbon = 1
+            if by_carbon:
+                carbon = sum(metabolite.n_carbon() for metabolite in rxn.reactants)
+            try:
+                flux += rxn.flux * carbon
+            except AssertionError:
+                # optimized flux can be out of bounds, in that case treat as zero flux
+                return 0
+        return flux
+
+    def carbon_yield(self):
+        """ mol product per mol carbon input
+
+        Returns
+        -------
+        float
+            the mol carbon atoms in the product (as defined by the model objective) divided by the mol carbon in the
+            input reactions (as defined by the model medium) """
+        try:
+            carbon_input_flux = self.total_flux(self.c_source_reactions)
+            carbon_output_flux = self.total_flux([self.product_reaction])
+            return carbon_output_flux / (carbon_input_flux * -1)
+        except (Infeasible, UndefinedSolution, ZeroDivisionError):
```

Comment:
UndefinedSolution is an error that should be raised. There is something going on with GLKP not returning a proper error if presolve is off, but UndefinedSolution should break. We should only trap infeasible solutions. The other types should be raise because they result from bad model/model constraints.